python/whylogs/core/view/segmented_dataset_profile_view.py

```diff
@@ -0,0 +1,184 @@
+import tempfile
+from logging import getLogger
+from typing import Any, Dict, Optional
+
+from whylogs.api.writer.writer import Writable
+from whylogs.core.proto import (
+    ChunkHeader,
+    ChunkMessage,
+    ChunkOffsets,
+    DatasetProfileHeader,
+    DatasetProperties,
+    DatasetSegmentHeader,
+    MetricComponentMessage,
+)
+from whylogs.core.proto import Segment as SegmentMessage
+from whylogs.core.segment import Segment
+from whylogs.core.segmentation_partition import SegmentationPartition
+from whylogs.core.utils import write_delimited_protobuf
+from whylogs.core.utils.timestamp_calculations import to_utc_milliseconds
+from whylogs.core.view.dataset_profile_view import (
+    WHYLOGS_MAGIC_HEADER_BYTES,
+    DatasetProfileView,
+)
+from whylogs.migration.converters import (
+    _generate_segment_tags_metadata,
+    v1_to_dataset_profile_message_v0,
+)
+
+logger = getLogger(__name__)
+
+
+class SegmentedDatasetProfileView(Writable):
+    _profile_view: DatasetProfileView
+    _segment: Segment
+    _partition: SegmentationPartition
+
+    def __init__(
+        self,
+        *,
+        profile_view: DatasetProfileView,
+        segment: Segment,
+        partition: SegmentationPartition,
+    ):
+        self._profile_view = profile_view
+        self._segment = segment
+        self._partition = partition
+
+    @property
+    def segment(self) -> Segment:
+        return self._segment
+
+    @property
+    def partition(self) -> SegmentationPartition:
+        return self._partition
+
+    @property
+    def profile_view(self) -> DatasetProfileView:
+        return self._profile_view
+
+    @property
+    def dataset_timestamp(self):
+        return self.profile_view.dataset_timestamp
+
+    @property
+    def creation_timestamp(self):
+        return self.profile_view.creation_timestamp
+
+    def get_default_path(self) -> str:
+        return f"profile_{self._profile_view.creation_timestamp}_{self._segment.parent_id}_{'_'.join(self._segment.key)}.bin"
+
+    def _write_as_v0_message(self, path: Optional[str] = None, **kwargs: Any) -> None:
+        message_v0 = v1_to_dataset_profile_message_v0(self.profile_view, self.segment, self.partition)
+        path = path or self.get_default_path()
+        with open(path, "w+b") as out_f:
+            write_delimited_protobuf(out_f, message_v0)
+            # out_f.write(message_v0.SerializeToString())
+
+    def _write_v1(self, path: Optional[str] = None, **kwargs: Any) -> None:
```

Comment:
This code looks the same as what's in DatasetProfileView, could this get refactored to pull the method out allowing for less duplication?

Reply:
I do see the bit of segments at the bottom. Doesn't need to be in this release, but I'd recommend us extracting some of these as private methods and using them in series instead of this long somewhat complicated method that has a lot of reuse. 

I do see this doesn't inherit from DatasetProfileView. Maybe it should and those can be extracted up, or having an utility for potions of it for both of them

Reply:
Fair point.

This is just a container, since I didn't want to modify dataset profile, or create circular references, I just wanted a way to bind the segments and the profile for later serialization. There is some duplication in the interest of reducing risk (so I don't have to modify the core unsegmented scenario too much in DatasetProfile), so the new behavior is isolated to this class.

We can factor out the common parts of the file format so there is less duplication, or proceed with this and iterate?

Reply:
I think we're on to proceed and iterate. Maybe we can make an issue for an easy code refactor when someone has bandwidth? :)